src/profilers/profile-serializer.ts

```diff
@@ -15,132 +15,278 @@
+
+// A stack of function IDs.
 type Stack = Array<number>;
+
+// A function which appends the sample form of entry to samples.
+type AppendEntryToSamples =
+    (entry: Entry, samples: perftools.profiles.Sample[]) => void;
+
+// Profile node and stack trace to that node.
 interface Entry {
+  node: ProfileNode;
+  stack: Stack;
+}
+
+// Heap profile node and stack trace to that node.
+interface AllocationEntry {
+  node: AllocationProfileNode;
+  stack: Stack;
+}
+
+// Time profile node and stack trace to that node.
+interface TimeEntry {
   node: TimeProfileNode;
   stack: Stack;
 }
 
+// Fields which uniquely describe a function.
+interface FunctionKey {
+  name: string;
+  scriptId: number;
+}
+
+// Fields which uniquely describe a location.
+interface LocationKey {
+  function: FunctionKey;
+  lineNumber: number;
+  columnNumber: number;
+}
+
 /**
- * Converts v8 Profile into profile with profile format used by Stackdriver
- * Profiler.
- *
- * @param prof - profile to be converted.
- * @param intervalMicros - average time (microseconds) between samples.
+ * Used to build string table and access strings and their ids within the table
+ * when serializing a profile.
  */
-export function serializeTimeProfile(
-    prof: TimeProfile, intervalMicros: number) {
-  let samples: Array<perftools.profiles.Sample> = [];
-  let locations: Array<perftools.profiles.Location> = [];
-  let functions: Array<perftools.profiles.Function> = [];
-  let locationMap: Map<number, perftools.profiles.Location> = new Map();
-  let functionMap: Map<number, perftools.profiles.Function> = new Map();
-  let strings = [''];
+class StringTable {
+  strings: string[];
+  stringsMap: Map<string, number>;
 
-  let sampleValueType = createSampleValueType();
-  let timeValueType = createTimeValueType();
+  constructor() {
+    this.strings = [];
+    this.stringsMap = new Map<string, number>();
+  }
 
-  serializeNode(prof.topDownRoot);
+  /**
+   * Returns index of str within the table. Also adds str to string table if
+   * str is not in the table already.
+   *
+   * @param str
+   */
+  getIndexOrAdd(str: string): number {
+    let loc = this.stringsMap.get(str);
+    if (loc !== undefined) {
+      return loc;
+    }
+    loc = this.strings.length;
+    this.stringsMap.set(str, loc);
+    this.strings.push(str);
+    return loc;
+  }
 
-  return {
-    sampleType: [sampleValueType, timeValueType],
-    sample: samples,
-    location: locations,
-    function: functions,
-    stringTable: strings,
-    // opt drop_frames
-    // opt keep_frames
-    timeNanos: 1000 * 1000 * prof.endTime,                         // Nanos
-    durationNanos: 1000 * 1000 * (prof.endTime - prof.startTime),  // Nanos
+  /**
+   * Returns value type for samples counts (type:samples, units:count), and
+   * adds strings used in this value type to the table.
+   */
+  createSampleValueType(): perftools.profiles.ValueType {
```

Comment:
I am not sure this function and two below should be a part of StringTable class. They only use the public interface of the class and don't seem to clearly belong by functionality, so should be elsewhere imo.